### PR TITLE
Adding multifunction support for nic adaptors while passthrough to guest

### PR DESCRIPTION
**Pull request by prudmirya**
Existing test case supports single function pass-through for nic and storage adaptor, now i am adding multifunction support for nic adaptors.
Signed-off-by: prudhvi <mprudhvi@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by prudmirya** [comment]
i close the pull request  #1136, and open a new one

File: libvirt/tests/cfg/libvirt_pci_passthrough.cfg [hunk 1]
```diff
@@ -13,14 +13,19 @@
             # device to guest. Then this network device
             # will be unavailable on host.
             # E.g: pci_0000_05_00_0
-            libvirt_pci_net_dev_label = "ENTER.YOUR.PCI.LABEL"
+            libvirt_pci_net_dev_label = "pci_0001_0c_00_0"
```
Comment (balamuruhans) [inline libvirt/tests/cfg/libvirt_pci_passthrough.cfg]: please change the values to default,

File: libvirt/tests/cfg/libvirt_pci_passthrough.cfg [hunk 2]
```diff
@@ -13,14 +13,19 @@
...
             # Please enter the ip what is used by the device
             # you are going to attach to guest.
-            libvirt_pci_net_ip = "ENTER.YOUR.IP"
+            libvirt_pci_net_ip = "9.47.66.2"
```
Comment (balamuruhans) [inline libvirt/tests/cfg/libvirt_pci_passthrough.cfg]: same here

**Comment by prudmirya** [comment]
[job.txt](https://github.com/autotest/tp-libvirt/files/1193889/job.txt)
log attached

File: libvirt/tests/cfg/libvirt_pci_passthrough.cfg [hunk 3]
```diff
@@ -13,14 +13,19 @@
...
             # Please enter a vailable ip from the net device.
             # We need to ping it after attaching pci device
             # to guest to verify this device works well in guest.
-            libvirt_pci_server_ip = "ENTER.YOUR.SERVER.IP"
+            libvirt_pci_server_ip = "9.47.66.41"
```
Comment (balamuruhans) [inline libvirt/tests/cfg/libvirt_pci_passthrough.cfg]: same here

File: libvirt/tests/cfg/libvirt_pci_passthrough.cfg [hunk 4]
```diff
@@ -13,14 +13,19 @@
...
+            libvirt_pci_net_dev_name = "Broadcom"
+            # Enter netmask in CIDR notation
+            libvirt_pci_net_mask = "24"
+            # if the adapter is multifunction please spcify as 'yes' else 'no'
+            libvirt_pci_multifunction = 'yes'
```
Comment (balamuruhans) [inline libvirt/tests/cfg/libvirt_pci_passthrough.cfg]: let it be no by default, let us configure when required.

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 5]
```diff
@@ -48,75 +49,90 @@ def run(test, params, env):
...
+        for val in pci_ids:
+            temp = val.replace(":", "_")
+            pci_devs.extend(["pci_"+temp])
+        count = 1
+        pci_id = re.sub('[:.]', '_', pci_id)
+        for val in pci_devs:
+            val = val.replace(".", "_")
+            if sriov:
+                if multifunction == 'yes':
+                    test.cancel("As of now multifunction with sriov not supported")
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: I think multifunction for sriov is not required.

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 6]
```diff
@@ -48,75 +49,90 @@ def run(test, params, env):
...
+                #          modprobe igb with max_vfs=7 failed.
+                # command '|| modprobe igb' is a handler of error occured
+                #          when we load igb again. If command 2 failed,
+                #          this command will be executed to recover network.
+                output_file = os.path.join(test.tmpdir, "output")
+                if os.path.exists(output_file):
+                    os.remove(output_file)
+                mod_cmd = ("modprobe -r igb && modprobe igb max_vfs=7 ||"
+                           "echo 'FAIL' > %s && modprobe igb &" % output_file)
+                result = utils.run(mod_cmd, ignore_status=True)
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: Test would probably break here as utils is removed from importing.

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 7]
```diff
@@ -48,75 +49,90 @@ def run(test, params, env):
...
+                if os.path.exists(output_file):
+                    test.error("Failed to modprobe igb with max_vfs=7.")
+                # Get the virtual function pci device which was generated above.
+                pci_xml = NodedevXML.new_from_dumpxml(pci_dev)
+                virt_functions = pci_xml.cap.virt_functions
+                if not virt_functions:
+                    test.fail("Init virtual function failed.")
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: this would be test error

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 8]
```diff
@@ -48,75 +49,90 @@ def run(test, params, env):
...
+                pci_address = virt_functions[0]
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: we have to handle for multiple VFs in each pci devs

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 9]
```diff
@@ -48,75 +49,90 @@ def run(test, params, env):
...
+                pci_dev = utils_test.libvirt.pci_label_from_address(pci_address,
+                                                                    radix=16)
+
+                # Find the network name (ethX) is using this pci device.
+                network_service = service.Factory.create_service("network")
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: for Ubuntu it will be `networking`

File: libvirt/tests/src/libvirt_pci_passthrough.py [hunk 10]
```diff
@@ -48,75 +49,90 @@ def run(test, params, env):
...
+                # command '|| echo 'FAIL' > output_file' is a flag to mean
+                #          modprobe igb with max_vfs=7 failed.
+                # command '|| modprobe igb' is a handler of error occured
+                #          when we load igb again. If command 2 failed,
+                #          this command will be executed to recover network.
+                output_file = os.path.join(test.tmpdir, "output")
+                if os.path.exists(output_file):
+                    os.remove(output_file)
+                mod_cmd = ("modprobe -r igb && modprobe igb max_vfs=7 ||"
+                           "echo 'FAIL' > %s && modprobe igb &" % output_file)
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough.py]: This sriov setup is handled from test_setup.PciAssignable()

**Comment by prudmirya** [comment]
create new PR #1215 , i am closing this current PR

**Pull request closed without merging**